internal/http/parser/http_response.go

```diff
@@ -0,0 +1,225 @@
+		}
+	}
+	statusCode := string(statusB)
+	if len(statusCode) != 3 {
+		reader.Seek(int64(-2), io.SeekCurrent)
+		return HTTPResponse{}, BadProtocolError{
+			Reader:  reader,
+			Message: "Invalid status-code field length, Expected: 3 digits, Received: " + strconv.Itoa(len(statusCode)),
+		}
+	}
+	intStatusCode, err := strconv.Atoi(statusCode)
+	if err != nil {
+		reader.Seek(int64(-2), io.SeekCurrent)
+		return HTTPResponse{}, BadProtocolError{
+			Reader:  reader,
+			Message: "Invalid status-code field, Expected integer value, Received: " + statusCode,
+		}
+	}
+	SL.StatusCode = intStatusCode
+	sectionsFound++
+
+	// Intentionally lax. rfc9112.html#section-4-8
+	reasonB, err := ReadUntilCRLF(reader)
+	if err == io.EOF {
+		return HTTPResponse{}, IncompleteInputError{
+			Reader:  reader,
+			Message: "Expected CRLF after status line",
+		}
+	}
+	reason := string(reasonB)
+	SL.Reason = reason
+	R.StatusLine = SL
+	sectionsFound++
+
+	for !allHeadersFound {
+		offsetBeforeCRLF := GetReaderOffset(reader)
+		possibleHeaderLine, err := ReadUntilCRLF(reader)
+		if err == io.EOF {
+			if len(possibleHeaderLine) == 0 {
+				return R, IncompleteInputError{
+					Reader:  reader,
+					Message: "Expected empty line after header section",
+				}
+			} else {
+				return R, IncompleteInputError{
+					Reader:  reader,
+					Message: "Expected CRLF after header value",
+				}
+			}
+		}
+		if len(possibleHeaderLine) == 0 {
+			// Headers finished
+			allHeadersFound = true
+			sectionsFound++
+		} else {
+			// We know header is present
+			reader.Seek(int64(offsetBeforeCRLF), io.SeekStart)
+			key, err := ReadUntil(reader, []byte(":"))
+			if err == io.EOF {
+				return R, IncompleteInputError{
+					Reader:  reader,
+					Message: "Expected ':' after header key",
+				}
+			}
+			if key[len(key)-1] == ' ' {
+				// No WS before separator
+				return R, BadProtocolError{
+					Reader:  reader,
+					Message: "No whitespace allowed before header separator",
+				}
+			}
+			value, err := ReadUntilCRLF(reader)
+			if err == io.EOF {
+				return R, IncompleteInputError{
+					Reader:  reader,
+					Message: "Expected CRLF after header value",
+				}
+			}
+			H := Header{
+				// 9110: 5.5-5: Replace CR, LF or NUL with SP
+				Key:   string(key),
+				Value: strings.TrimSpace(string(ReplaceCharsWithSpace(value, [][]byte{CR, LF, NUL}))),
+			}
+			R.Headers = append(R.Headers, H)
+		}
+	}
+
+	// Content is present
+	if R.ContentLength() != -1 {
+		content, err := ReadBytes(reader, R.ContentLength())
+		if err == io.EOF {
+			return R, IncompleteInputError{
+				Reader:  reader,
+				Message: fmt.Sprintf("Expected content of length %d, Received content of length %d", R.ContentLength(), len(content)),
+			}
+		}
+		R.Body = content
+	}
+	sectionsFound++
+
+	if sectionsFound != 5 {
+		return R, BadProtocolError{
+			Reader:  reader,
+			Message: "Expected 5 sections in response",
```

Comment:
@ryan-gang 🤦 How is a user supposed know what a "section" is? 

<img width="855" alt="Screenshot 2024-05-04 at 23 09 03" src="https://github.com/codecrafters-io/http-server-tester/assets/3893573/1e22f377-dde0-4382-97a8-4bea6e2af259">